### PR TITLE
chore: add JS only higher order functions lab to frontend cert

### LINKED ISSUE #519: chore: add JS only higher order functions lab to frontend cert
We need a lab for campers to practice working with higher order functions like `map`, `filter`, etc.
This will be JS only. Not HTML or CSS or DOM manipulation.

Here is the workshop they will do before this
https://github.com/freeCodeCamp/CurriculumExpansion/blob/main/frontend-cert/js-projects/library-manager/script.js

They will also know all of the fundamentals (strings, functions, objects, arrays, etc)

This doesn't need to cover everything in the workshop, just pick a couple of things like `map` or `filter`. 

## Acceptance criteria

- [ ] create a new folder of the project in the `frontend-cert/js-projects` directory
- [ ] add your completed project in that new folder
- [ ] create a markdown file for your proposed user stories. You should call it `user-stories.md` and place it in your new project folder

### PR DESCRIPTION
**Pull request by fhsinchy**
Checklist:

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a [descriptive title](https://contribute.freecodecamp.org/#/how-to-open-a-pull-request?id=prepare-a-good-pr-title) (**not** a vague title like `Update index.md`)

Closes #519

### CODE REVIEW

File: frontend-cert/js-projects/higher-order-functions/script.js [hunk 1]
```diff
@@ -0,0 +1,48 @@
+function sortByYear(book1, book2) {
+  if (book1.release_year < book2.release_year) {
+    return -1;
+  }
+
+  if (book1.release_year < book2.release_year) {
```
Comment (zairahira) [inline frontend-cert/js-projects/higher-order-functions/script.js:6]: To match the instruction:

```suggestion
  if (book1.release_year > book2.release_year) {
```

File: frontend-cert/js-projects/higher-order-functions/user-stories.md [hunk 2]
```diff
@@ -0,0 +1,15 @@
+You should have an array of objects named `books` where each object in the array should have a string `title`, another string `author_name`, and a number `release_year`.
+
+You should call the filter higher-order function on the `books` array with a callback to filter out books written after a certain year such as 1950 and save the filtered array in a new array named `filteredBooks`.
```
Comment (zairahira) [inline frontend-cert/js-projects/higher-order-functions/user-stories.md:3]:
```suggestion
You should call the `filter` higher-order function on the `books` array with a callback to filter out books written after a certain year such as 1950 and save the filtered array in a new array named `filteredBooks`.
```

File: frontend-cert/js-projects/higher-order-functions/user-stories.md [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+
+You should call the `sort` higher order function on the `filteredBooks` array to filter the books according to their `release_year` in ascending order and save the filtered array in a new array named `filteredBooksSorted`.
```
Comment (zairahira) [inline frontend-cert/js-projects/higher-order-functions/user-stories.md:5]:
```suggestion
You should call the `sort` higher order function by passing the `sortByYear` callback function on the `filteredBooks` array to sort the books according to their `release_year` in ascending order and save the sorted array in a new array named `filteredBooksSorted`.
```

**Review by zairahira — CHANGES_REQUESTED** [review]
This is a good practice project 👍️

I feel the instructions should be ordered like this because in one of the steps(updated step 3) the `sortByYear` is used before it is has been defined:


```md
You should have a callback function named `sortByYear` that accepts two books, `book1`, and `book2` as parameter for sorting the array.

The `sortByYear` function should return `-1` if the `release_year` of `book1` object is smaller than that of the `book2` object.

The `sortByYear` function should return `1` if the `release_year` of `book1` object is bigger than that of the `book2` object.

The `sortByYear` function should return `0` for all other scenarios.

You should have an array of objects named `books` where each object in the array should have a string `title`, another string `author_name`, and a number `release_year`.

You should call the `filter` higher-order function on the `books` array with a callback to filter out books written after a certain year such as 1950 and save the filtered array in a new array named `filteredBooks`.

You should call the `sort` higher order function by passing the `sortByYear` callback function
on the `filteredBooks` array to sort the books according to their `release_year` in ascending order and save the sorted array in a new array named `filteredBooksSorted`. 

You should call the `forEach` higher order function on the `filteredBooksSorted` array and print out the details of each book following the `<BOOK_TITLE> by <BOOK_AUTHOR> (<RELEASE_YEAR>)` format.

```

**Comment by fhsinchy** [comment]
@zairahira I have implemented the suggestion and now it's ready for another review.

File: frontend-cert/js-projects/higher-order-functions/user-stories.md [hunk 4]
```diff
@@ -0,0 +1,15 @@
+You should have a callback function named `sortByYear` that accepts two books, `book1`, and `book2` as parameter for sorting the array.
+
+The `sortByYear` function should return `-1` if the `release_year` of `book1` object is smaller than that of the `book2` object.
+
+The `sortByYear` function should return `1` if the `release_year` of `book1` object is bigger than that of the `book2` object.
+
+The `sortByYear` function should return `0` for all other scenarios.
+
+You should have an array of objects named `books` where each object in the array should have a string `title`, another string `author_name`, and a number `release_year`.
```
Comment (zairahira) [inline frontend-cert/js-projects/higher-order-functions/user-stories.md:9]: Maybe we can mention that "You should add at least 3 items to your array".

**Review by zairahira — APPROVED** [review]
Left one small suggestion, rest is good to go 👍️

**Review by jdwilkin4 — APPROVED** [review]
I went ahead and pushed up a change to include that last review comment.

This is good to go in

**Pull request merged**